### PR TITLE
[ONNX] Align behavior of ReduceL2-11, 13, 18 with original framework

### LINKED ISSUE #20560: [Good First Issue]: Align behavior of ONNX Frontend function ReduceL2-11, 13, 18 with original framework
### Context

Neural networks are graphs consisting of nodes called operators. Each operator corresponds to a mathematical function, usually described in framework's documentation or an AI standard, such as ONNX.
OpenVINO ONNX Frontend is a component responsible for working with ONNX graphs and requires implementation of different ONNX operators in order to use ONNX models.
This task requires alignment between OpenVINO ONNX Frontend and original framework implementations of ReduceL2 for next list of opsets: [opset 11](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#ReduceL2-11),  [opset 13](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#ReduceL2-13),  [opset 18](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#ReduceL2-18)
Necessary help will be provided by ONNX Fronted team.

### What needs to be done?

First of all, please, take a look on [ReduceMax](https://github.com/openvinotoolkit/openvino/pull/23475) PR for a reference.

Operator details can be found in [ONNX Operators](https://github.com/onnx/onnx/blob/main/docs/Operators.md#ReduceL2)
More details can be found in ONNX Changelog: [opset 11](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#ReduceL2-11),  [opset 13](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#ReduceL2-13),  [opset 18](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#ReduceL2-18)
1. Function already has a common implementation in OpenVINO. First of all, you need to review a documentation and prepare a table with differences between versions. It could be, for instance, a missing property, extended/reduced coverage of existing property, etc...
2. Copy existing implementation [here](https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/onnx/frontend/src/op) to make it aligned with original framework (extend or reduce coverage of a common implementation). Copy of modified implementation should be in a defined opset, or in opset 1 in case it implements oldest implementation. [Example of multi-opset operation](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/softmax.cpp#L50).
3. Register the function in [ops_bridge.cpp](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/ops_bridge.cpp#L554-L560) while keeping alphabetical order
4. Create test model(s) in [ONNX models directory](https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/onnx/tests/models/). OpenVINO test infrastructure then converts prototxt files to ONNX models - you will use those models later in tests
5. Add tests covering all use cases [here](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/tests/onnx_import.in.cpp)
6. Check [Python xfailed tests](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/tests/tests_python/test_backend.py) to find a test marked as a xfailed for added functionality. If any exist - remove corresponding lines and try to verify by using cmdline "python -m pytest -k name_of_test".
More details in [adding operators to ONNX Frontend guide](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/how_to_add_op.md)

### Example Pull Requests

_No response_

### Resources

- [Contribution guide - start here!](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [Intel DevHub Discord channel](https://discord.gg/wPuqAujS) - engage in discussions, ask questions and talk to OpenVINO developers
- [What is OpenVINO?](https://github.com/openvinotoolkit/openvino#what-is-openvino-toolkit)
- [User documentation](https://docs.openvino.ai/)
- Guides on [ONNX Frontend architecture](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/architecture.md) and [ONNX Frontend tests](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/tests.md) are also available
- You can also see a [blog post on contributing to OpenVINO](https://medium.com/openvino-toolkit/how-to-contribute-to-an-ai-open-source-project-c741f48e009e)

### Contact points

@gkrivor 

### Ticket

_No response_

### PR DESCRIPTION
**Pull request by LucaTamSapienza**
### Details:
- I've aligned the ReduceL2 operation with opset 11, 13, and 18. I have some doubts about how to implement support for the tensor bfloat16 for opset 13 and also some doubts about opset 18. I've registered the function inside the ops_bridge.cpp, created test models, and added them inside onnx_import.in.cpp.
### Tickets:
 - Closes #20560

### CODE REVIEW

**Comment by p-wysocki** [comment]
Hello @LucaTamSapienza, is this PR ready for review?

**Comment by LucaTamSapienza** [comment]
> Hello @LucaTamSapienza, is this PR ready for review?

Hi @p-wysocki. The code compiles, but I have some pending issues. I have described my problems in the PR specification. I was waiting for someone to provide me with some clarifications on the doubts I encountered. I took a look at [softmax.cpp](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/softmax.cpp) file and to [this](https://github.com/openvinotoolkit/openvino/issues/20563) conversation to see how to implement the features of the different versions. Unfortunately, I still have some doubts about `opset13`, on how to support the `bfloat` input. And I'm not too sure about my implementation of `opset11`. If you could take a look at my code and give me some feedback, it would be greatly appreciated. Thanks!

File: src/frontends/onnx/frontend/src/op/reduce.cpp [hunk 1]
```diff
@@ -93,24 +93,28 @@ std::shared_ptr<ov::Node> get_reduction_axes_from_attr(const Node& node) {
...
-                                               bool axes_as_attr = true) {
+                                               bool axes_as_attr = true,
+                                               const std::pair<int64_t, int64_t>& axes_range = {
+                                               -std::numeric_limits<int64_t>::max(),
+                                               std::numeric_limits<int64_t>::max()}) {
     const std::int64_t keepdims = node.get_attribute_value<std::int64_t>("keepdims", 1);
 
     const auto reduction_axes = axes_as_attr ? get_reduction_axes_from_attr(node) : get_reduction_axes_from_input(node);
     if (reduction_axes != nullptr) {
+        // Check that axes are within the specified range
```
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/reduce.cpp:104]: I think you may not need such overcomplication after rebasing to a fresh master.

File: src/frontends/onnx/frontend/src/op/reduce.cpp [hunk 2]
```diff
@@ -158,6 +162,30 @@ ov::OutputVector reduce_sum_square(const ov::frontend::onnx::Node& node) {
 }
 
 }  // namespace set_1
+
+namespace set_11 {
+ov::OutputVector reduce_l2(const Node& node) {
+    if (node.get_ov_inputs().at(0).get_partial_shape().rank().is_static()) {
```
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/reduce.cpp:168]: Not sure it is needed. I suggest to add some missing checks for types. Checks should be aligned with original ONNX documentation.

File: src/frontends/onnx/frontend/src/op/reduce.cpp [hunk 3]
```diff
@@ -158,6 +162,30 @@ ov::OutputVector reduce_sum_square(const ov::frontend::onnx::Node& node) {
...
+
+namespace set_13 {
+ov::OutputVector reduce_sum(const ov::frontend::onnx::Node& node) {
+    return {make_ng_reduction_op<v1::ReduceSum>(node, node.get_ov_inputs().at(0), false)};
+}
+}  // namespace set_13
+
+namespace set_18 {
+ov::OutputVector reduce_l2(const ov::frontend::onnx::Node& node) {
+    return {make_ng_reduction_op<v4::ReduceL2>(node, node.get_ov_inputs().at(0))};
```
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/reduce.cpp:186]: Overall - good, just add missing checks

**Review by gkrivor — CHANGES_REQUESTED** [review]
Hi @LucaTamSapienza, thanks for your contribution!

Please, rebase your solution to a fresh master, because I've merged a change https://github.com/openvinotoolkit/openvino/pull/22462 with fix for all Reduce* operations which may affect your test cases.

Also, could you enable tests here https://github.com/openvinotoolkit/openvino/blob/96b72d0c823a44cab1ad89d9e6ca0d0e3ac75c0c/src/frontends/onnx/tests/tests_python/test_backend.py#L485

They should work after rebasing to a fresh master

**Comment by LucaTamSapienza** [comment]
> Please, rebase your solution to a fresh master, because I've merged a change #22462 with fix for all Reduce* operations which may affect your test cases.

Sure, i'll sync && pull immediatly.

> Also, could you enable tests here

Do you mean for v 11, 13 ,18?

File: src/frontends/onnx/frontend/src/op/reduce.cpp (hunk 1 above)
Comment (LucaTamSapienza) [reply src/frontends/onnx/frontend/src/op/reduce.cpp:104]: > I think you may not need such overcomplication after rebasing to a fresh master.

Ok, thanks for the advice!

File: src/frontends/onnx/frontend/src/op/reduce.cpp (hunk 3 above)
Comment (LucaTamSapienza) [reply src/frontends/onnx/frontend/src/op/reduce.cpp:173]: > missing checks

I'm checking [ONNX documentation](https://onnx.ai/onnx/operators/onnx__ReduceL2.html#reducel2) to understand what additional checks I should perform. I don't understand which ones you're referring to. Could you please be more specific about it?

**Comment by gkrivor** [comment]
Hi, I refactored reduce.cpp and reduce.hpp. I think it should be more simple to add a new functionality.

Please, rebase after merge.

https://github.com/openvinotoolkit/openvino/pull/23429

**Comment by LucaTamSapienza** [comment]
> Hi, I refactored reduce.cpp and reduce.hpp. I think it should be more simple to add a new functionality.
> 
> Please, rebase after merge.
> 
> #23429

Thanks for it, I see that it has been merged. I'll work on it.

**Comment by gkrivor** [comment]
Hi @LucaTamSapienza ,

Please take a look on https://github.com/openvinotoolkit/openvino/pull/23475 as a reference.

Especially take a look on test cases. For example, as I can see you've missed in opset-18 prototxt mention of an "axes" input. It is very important to verify this way.

Also check changes in onnx/tests/__init__.py and onnx/tests/tests_python/test_backend.py

**Comment by LucaTamSapienza** [comment]
> Please take a look on #23475 as a reference.

@gkrivor ,

Thank you very much for your help and linking me to this PR, I have properly aligned versions 13 and 18. I noticed that you left a comment under version 11. Perhaps I should remove the call from reduce.cpp file and simply use the same implementation used in v1 by calling it in reduce.hpp within v11? Something like 
```
namespace set_11 {
using set_1::reduce_l2;
} // namespace set_11
```



> Especially take a look on test cases. For example, as I can see you've missed in opset-18 prototxt mention of an "axes" input. It is very important to verify this way.

For this correction, I took inspiration from the prototxt of the `reduceMax` operation in `set_18`. Now should be good.

> Also check changes in onnx/tests/**init**.py and onnx/tests/tests_python/test_backend.py

I've deleted the ReduceL2 references from ticket 99962 && 99968 inside the __init__.py and removed for 99968 declaration about reduce_l2. I just have some doubts about this `"OnnxBackendNodeModelTest.test_reduce_l2_empty_set_cpu"`, should I leave this?

**Comment by gkrivor** [comment]
Hi @LucaTamSapienza, thanks for the update!

Regarding opset-11. Actually, I think it is a good idea, I also thought about something like this. But currently, we have an internal discussion about "following the standards". I suggest to keep it as is. When we will decide how strictly we want to support standards - I could easily extend/reduce code base.

Also, I don't recommend to touch this place because several developers are working on reduce* operations right now and you may introduce additional merge-conflicts between you and them.

About  "OnnxBackendNodeModelTest.test_reduce_l2_empty_set_cpu" - depends on a test result. If it become passing (you will see in test results), and leave it if it is still xfailed.

**Comment by LucaTamSapienza** [comment]
> Also, I don't recommend to touch this place because several developers are working on reduce* operations right now and you may introduce additional merge-conflicts between you and them.

You're right @gkrivor , I've just resolved some merge conflicts I had, now everything should be fine.

> About "OnnxBackendNodeModelTest.test_reduce_l2_empty_set_cpu" - depends on a test result. If it become passing (you will see in test results), and leave it if it is still xfailed.

I ran `/bin/arm64/Release/ov_onnx_frontend_tests --gtest_filter=\*reduce_l2\*` and all the tests passed successfully.

File: src/frontends/onnx/frontend/src/op/reduce.cpp [hunk 4]
```diff
@@ -186,25 +186,28 @@ ov::OutputVector reduce_sum_square(const ov::frontend::onnx::Node& node) {
...
 
-/*
-    Opset 11 is skipped because there are no significant difference between opset1 and opset 11.
-    Found difference is:
-    1. Operations (except ReduceMin and ReduceMax) are lost mention of zero-rank input behavior
-       from their description. We assume it shouldn't be worse than opset 1.
-    2. Opset 11 introduced requirement for axes values to be in a range [-r, r-1] where r = rank(data)
-       Same time Reduce* operations in OpenVINO has same requirement from first version
-*/
+namespace set_11 {
```
Comment (gkrivor) [inline src/frontends/onnx/frontend/src/op/reduce.cpp:189]: Please, do not introduce opset-11 right now, everything else looks good

Comment (LucaTamSapienza) [reply src/frontends/onnx/frontend/src/op/reduce.cpp]: > Please, do not introduce opset-11 right now, everything else looks good

Sure, I'll remove it also from the header file

**Comment by gkrivor** [comment]
build_jenkins

**Review by gkrivor — APPROVED** [review]

**Comment by gkrivor** [comment]
build_jenkins

**Comment by gkrivor** [comment]
build_jenkins
